src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java

```diff
@@ -0,0 +1,976 @@
+    private static final String CLASS_DATA_DESC = methodType(Object.class, MethodHandles.Lookup.class, String.class, Class.class).descriptorString();
+    private static final String RELEASE0_DESC = methodType(void.class).descriptorString();
+    private static final String ACQUIRE0_DESC = methodType(void.class).descriptorString();
+
+    private static final Handle BSM_CLASS_DATA = new Handle(
+            H_INVOKESTATIC,
+            METHOD_HANDLES_INTRN,
+            "classData",
+            CLASS_DATA_DESC,
+            false);
+    private static final ConstantDynamic CLASS_DATA_CONDY = new ConstantDynamic(
+            ConstantDescs.DEFAULT_NAME,
+            OBJECT_DESC,
+            BSM_CLASS_DATA);
+
+    private static final String CLASS_NAME_DOWNCALL = "jdk/internal/foreign/abi/DowncallStub";
+    private static final String CLASS_NAME_UPCALL = "jdk/internal/foreign/abi/UpcallStub";
+    private static final String METHOD_NAME = "invoke";
+
+    private static final String SUPER_NAME = OBJECT_INTRN;
+
+    private static final SoftReferenceCache<FunctionDescriptor, MethodHandle> UPCALL_WRAPPER_CACHE = new SoftReferenceCache<>();
+
+    // Instance fields start here
+    private final MethodVisitor mv;
+    private final MethodType callerMethodType;
+    private final CallingSequence callingSequence;
+    private final ABIDescriptor abi;
+    private final MethodType leafType;
+
+    private int localIdx = 0;
+    private int[] paramIndex2ParamSlot;
+    private int[] leafArgSlots;
+    private int[] scopeSlots;
+    private int curScopeLocalIdx = -1;
+    private int RETURN_ALLOCATOR_IDX = -1;
+    private int CONTEXT_IDX = -1;
+    private int RETURN_BUFFER_IDX = -1;
+    private int RET_VAL_IDX = -1;
+    private Deque<Class<?>> typeStack;
+    private List<Class<?>> leafArgTypes;
+    private int paramIndex;
+    private long retBufOffset; // for needsReturnBuffer
+
+    private BindingSpecializer(MethodVisitor mv, MethodType callerMethodType, CallingSequence callingSequence, ABIDescriptor abi, MethodType leafType) {
+        this.mv = mv;
+        this.callerMethodType = callerMethodType;
+        this.callingSequence = callingSequence;
+        this.abi = abi;
+        this.leafType = leafType;
+    }
+
+    static MethodHandle specialize(MethodHandle leafHandle, CallingSequence callingSequence, ABIDescriptor abi) {
+        if (callingSequence.forUpcall()) {
+            MethodHandle wrapper = UPCALL_WRAPPER_CACHE.get(callingSequence.functionDesc(), fd -> doSpecialize(leafHandle, callingSequence, abi));
+            return MethodHandles.insertArguments(wrapper, 0, leafHandle); // lazily customized for leaf handle instances
+        } else {
+            return doSpecialize(leafHandle, callingSequence, abi);
+        }
+    }
+
+    static MethodHandle doSpecialize(MethodHandle leafHandle, CallingSequence callingSequence, ABIDescriptor abi) {
```

Comment:
I think thise method should be split in to, one version for the upcall, one for the downcall, i do not see the point of merging the two codes.

Reply:
I've split this method into a specializeDowncall and specializeUpcall method.